lib/src/main/java/oiccli/Service.java

```diff
@@ -0,0 +1,443 @@
+        Map<String, Object> arArgs = new HashMap<>(args);
+        Object value;
+        for (String property : this.msgType.cParam.keySet()) {
+            if (!arArgs.containsKey(property)) {
+                value = clientInfo.getClass().getField(property).get(property);
+                if (value != null) {
+                    arArgs.put(property, value);
+                } else {
+                    arArgs.put(property, this.defaultRequestArgs.get(property));
+                }
+            }
+        }
+
+        return arArgs;
+    }
+
+    public List<Map<String, Object>> doPreConstruct(ClientInfo clientInfo, Map<String, Object> requestArgs, Map<String, String> args) {
+        Map<String, Object> postArgs = new HashMap<>();
+        for (Object method : this.preConstruct) {
+                /*request_args, _post_args = meth(cli_info, request_args, **kwargs)
+                post_args.update(_post_args)*/
+        }
+
+        return Arrays.asList(requestArgs, postArgs);
+    }
+
+    public Message doPostConstruct(ClientInfo clientInfo, Map<String, Object> requestArgs, Map<String, Object> postArgs) {
+
+    }
+
+    public void doPostParseResponse(Message response, ClientInfo clientInfo, String state, Map<String, KeyJar> args) {
+        /*
+          for meth in self.post_parse_response:
+            meth(resp, cli_info, state=state, **kwargs)
+         */
+    }
+
+    public Message constructMessage(ClientInfo clientInfo, Map<String, Object> requestArgs, Map<String, String> args) throws NoSuchFieldException, IllegalAccessException {
+
+        if (requestArgs == null) {
+            requestArgs = new HashMap<>();
+        }
+
+        List<Map<String, Object>> returnedArgs = this.doPreConstruct(clientInfo, requestArgs, args);
+
+        if (!this.msgType.c_param.containsKey("state")) {
+            args.remove("state");
+        }
+
+        Map<String, Object> argsParam = this.gatherRequestArgs(clientInfo, requestArgs);
+        this.msgType(argsParam);
+
+        return this.doPostConstruct(clientInfo, requestArgs, returnedArgs.get(1));
+    }
+
+    public String getEndpoint(Map<String, String> args) throws MissingEndpoint {
+        String uri = args.get("endpoint");
+
+        if (uri != null) {
+            args.remove("endpoint");
+        } else {
+            if (StringUtil.isNotNullAndNotEmpty(this.endpoint)) {
+                uri = this.endpoint;
+            } else {
+                throw new MissingEndpoint("No endpoint specified");
+            }
+        }
+
+        return uri;
+    }
+
+    public Map<String, Object> uriAndBody(Message cis, String method, Map<String, String> args) throws UnsupportedEncodingException, UnsupportedType, MissingEndpoint {
+        String uri = this.getEndpoint(args);
+        Map<String, Object> response = Util.getOrPost(uri, method, cis, args);
```

Comment:
rather than `getOrPost`, consider having a network stack or at least a class to handle them. Whitelist values for the methods, using things like enums.